# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/ActorsParent.cpp
# Commit: 2fb182664097
# Full Hash: 2fb182664097f1b22021eec844d0d1adbffc40dc
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-05 09:38:58
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Support multiple variants of hasResult expectations in CreateAndExecuteSingleStepStatement. r=dom-workers-and-storage-reviewers,ttung
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D97461
# ==============================================================================

diff -r 2ed5e139d784 -r 2fb182664097 dom/quota/ActorsParent.cpp
--- a/dom/quota/ActorsParent.cpp	Fri Dec 04 14:56:18 2020 +0000
+++ b/dom/quota/ActorsParent.cpp	Fri Dec 04 15:10:07 2020 +0000
@@ -344,25 +344,15 @@
 Result<int32_t, nsresult> LoadCacheVersion(mozIStorageConnection& aConnection) {
   AssertIsOnIOThread();
 
-  nsCOMPtr<mozIStorageStatement> stmt;
-  nsresult rv = aConnection.CreateStatement(
-      "SELECT cache_version FROM database"_ns, getter_AddRefs(stmt));
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-
-  bool hasResult;
-  rv = stmt->ExecuteStep(&hasResult);
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-
-  if (NS_WARN_IF(!hasResult)) {
-    return Err(NS_ERROR_FILE_CORRUPTED);
-  }
+  QM_TRY_INSPECT(const auto& stmt,
+                 CreateAndExecuteSingleStepStatement<
+                     SingleStepResult::ReturnNullIfNoResult>(
+                     aConnection, "SELECT cache_version FROM database"_ns));
+
+  QM_TRY(OkIf(stmt), Err(NS_ERROR_FILE_CORRUPTED));
 
   int32_t version;
-  rv = stmt->GetInt32(0, &version);
+  nsresult rv = stmt->GetInt32(0, &version);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return Err(rv);
   }
@@ -657,25 +647,15 @@
     mozIStorageConnection& aConnection) {
   AssertIsOnIOThread();
 
-  nsCOMPtr<mozIStorageStatement> stmt;
-  nsresult rv = aConnection.CreateStatement("SELECT version FROM database"_ns,
-                                            getter_AddRefs(stmt));
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-
-  bool hasResult;
-  rv = stmt->ExecuteStep(&hasResult);
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-
-  if (NS_WARN_IF(!hasResult)) {
-    return Err(NS_ERROR_FILE_CORRUPTED);
-  }
+  QM_TRY_INSPECT(const auto& stmt,
+                 CreateAndExecuteSingleStepStatement<
+                     SingleStepResult::ReturnNullIfNoResult>(
+                     aConnection, "SELECT version FROM database"_ns));
+
+  QM_TRY(OkIf(stmt), Err(NS_ERROR_FILE_CORRUPTED));
 
   int32_t version;
-  rv = stmt->GetInt32(0, &version);
+  nsresult rv = stmt->GetInt32(0, &version);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return Err(rv);
   }
@@ -4522,24 +4502,13 @@
   bool loadQuotaFromCache = false;
 
   if (mCacheUsable) {
-    nsCOMPtr<mozIStorageStatement> stmt;
-    rv = mStorageConnection->CreateStatement(
-        nsLiteralCString("SELECT valid, build_id "
-                         "FROM cache"),
-        getter_AddRefs(stmt));
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return rv;
-    }
-
-    bool hasResult;
-    rv = stmt->ExecuteStep(&hasResult);
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return rv;
-    }
-
-    if (NS_WARN_IF(!hasResult)) {
-      return NS_ERROR_FILE_CORRUPTED;
-    }
+    QM_TRY_INSPECT(
+        const auto& stmt,
+        CreateAndExecuteSingleStepStatement<
+            SingleStepResult::ReturnNullIfNoResult>(
+            *mStorageConnection, "SELECT valid, build_id FROM cache"_ns));
+
+    QM_TRY(OkIf(stmt), Err(NS_ERROR_FILE_CORRUPTED));
 
     int32_t valid;
     rv = stmt->GetInt32(0, &valid);